okhttp/src/main/java/okhttp3/internal/ws/WebSocketExtensions.kt

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.ws
+
+import java.io.IOException
+import okhttp3.Response
+import okhttp3.internal.delimiterOffset
+import okhttp3.internal.trimSubstring
+
+/**
+ * Models the contents of a `Sec-WebSocket-Extensions` response header. OkHttp honors one extension
```

Comment:
@swankjesse , from what I see, the PR is aligned with the 1..3 from above, so everything is fine here.

Reply:
We need to track the server's behavior for context takeover by making reset() calls at appropriate times. Otherwise we'll be decompressing with the wrong compression context. So we cannot ignore the server’s context takeover configuration.

Reply:
(I think your #3 should be MUST, not MAY.)

Reply:
Re-read the source and specs. You're right, we need to support `server_no_context_takeover` in response, and apply it to MessageDeflater (eg for client->server stream part). 

For some reason I missed the part that we cannot make context_takeover asymmetrical, as it applies to both reading and writing simultaneously on each party. 

Fixed the comment above, thanks.

Reply:
@swankjesse , please see above, I had to update the comment.

Reply:
However, it might be the case that I miss the point and we have to add the `contextTakeover` parameter to the MessageInflater too, and reset the `inflater` on every message read.

Reply:
I tested this experimentally. If the reset calls don’t align between deflater and inflater I get an exception:

https://gist.github.com/swankjesse/b504eb456474e4c5f47fbbeba1ef0244#file-resettest-kt-L69

Reply:
It fails indeed.
Thanks for double checking! Turns out I was under incorrect impression on how it works.

We'll have to add `contextTakeover` parameter based on `server_no_context_takeover` option from server response then.